### ISSUE TITLE
Development option to drop if any changeset is modified

### ISSUE THREAD

**Issue by mjeffrey** [liquibase/liquibase#4347]
### Search first

- [X] I searched and no similar issues were found

### Rationale

During (early) development changesets are often modified and it would be great to just be able to recreate the DB if one was changed (not systematically like dropFirst).

Some teams set dropFirst in their project for local builds but this is pretty annoying when you want the database to survive restarts. It is a bit of a blunt instrument.

The downside of this setting is that there is a risk it is left on for other environments (it should really only be used locally, dropFirst is a bit more obvious here).

Despite this it would be great to give teams this option as well.

### Proposal

If you are interested in this option I am willing to create a PR for it and also for the Spring Boot starter once it is merged.
I was planning on something like this (not tested) in the `DatabaseUpdateTask` and would be interested in your feedback. If you are OK with it I'd do the tests and docs and submit a PR.

```
@Override
public void executeWithLiquibaseClassloader() throws BuildException {
    Writer writer = null;
    Liquibase liquibase = getLiquibase();
    try {
        FileResource outputFile = getOutputFile();
        if(outputFile != null) {
            writer = getOutputFileWriter();
            liquibase.update(toTag, new Contexts(getContexts()), getLabelFilter(), writer);
        } else {
            if (dropFirst) {
                liquibase.dropAll();
            } else if (dropIfChanged) {
                dropIfChanged(liquibase);
            }
            liquibase.update(toTag, new Contexts(getContexts()), getLabelFilter());
        }
    } catch (LiquibaseException e) {
        throw new BuildException("Unable to update database: " + e.getMessage(), e);
    } catch (UnsupportedEncodingException e) {
        throw new BuildException("Unable to generate update SQL. Encoding [" + getOutputEncoding() + "] is not supported.", e);
    } catch (IOException e) {
        throw new BuildException("Unable to generate update SQL. Error creating output writer.", e);
    } finally {
        FileUtils.close(writer);
    }
}

private void dropIfChanged(Liquibase liquibase) throws LiquibaseException {
    try {
        liquibase.getChangeSetStatuses(new Contexts(getContexts()), getLabelFilter());
    } catch (ValidationFailedException validationFailedException) {
        if (!validationFailedException.getInvalidMD5Sums().isEmpty()) {
            liquibase.dropAll();
        } else {
            throw validationFailedException;
        }
    }
}
```

### Steps To Reproduce

Not applicable - feature request

### Expected/Desired Behavior

Not applicable - feature request

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

**Comment by nvoxland** [comment]
That is a good idea for a feature. I could see it helping with the inner development loop when first adding features.

We're trying to have logic be more consistent across all the integrations, and I could see this being just as useful in the CLI etc. so finding a good spot to make the change in liquibase-core is probably best. Probably in `AbstractUpdateCommandStep`? 

Rather than dropping on change, calling the rollback logic will be generically useful. That logic is already there and will handle things like renameColumn etc. too.

For safety, it's probably best to only allow rollback of the last changeset(s) executed and not allow rollback of ones before a non-changed one? So anything before the last unchanged changeset only throws validation errors on checksum changes like they currently do. That would keep it more isolated to the actively-under-development changesets and also won't cause problem where the later changesets are impacting what can be rolled back safely.

` there is a risk it is left on for other environments` Yes there is, but all we can do is trust that the user knows what they are doing. Adding a CommandArgumentDefinition to the update commands to optionally enable the new behavior lets them control it in a variety of ways so they can ensure it's only enabled in the correct environments.

**Comment by filipelautert** [comment]
We're currently triaging our backlog to ensure we can focus on the issues that matter most. With a large number of open issues, it becomes difficult to give each one the attention it deserves.

This issue has been closed due to inactivity — there have been no updates for over 30 days. If this is still relevant to you, please feel free to reopen it or create a new issue with updated context. We appreciate your contribution!
